hl7/parser.py

```diff
@@ -35,16 +91,213 @@ def parse(line, encoding="utf-8", factory=Factory):
+        ],
+    }
+    # If the BHS/BTS were present, use those to set up the batch
+    # otherwise default
+    if batch:
+        batch = parse(batch, encoding=encoding, factory=factory)
+        kwargs["esc"] = batch.esc
+        kwargs["separators"] = batch.separators
+        kwargs["factory"] = batch.factory
+    parsed = factory.create_batch(**kwargs)
+    # If the BHS/BTS were present then set them
+    if batch:
+        parsed.header = batch.segment("BHS")
+        try:
+            parsed.trailer = batch.segment("BTS")
+        except KeyError:
+            logger.warning("Missing BTS segment")
+    return parsed
+
+
+def parse_batch(lines, encoding="utf-8", factory=Factory):
+    """Returns a instance of a :py:class:`hl7.Batch`
+    that allows indexed access to the messages.
+
+    A custom :py:class:`hl7.Factory` subclass can be passed in to be used when
+    constructing the batch and it's components.
+
+    .. note::
+
+        HL7 usually contains only ASCII, but can use other character
+        sets (HL7 Standards Document, Section 1.7.1), however as of v2.8,
+        UTF-8 is the preferred character set [#]_.
+
+        python-hl7 works on Python unicode strings. :py:func:`hl7.parse_batch`
+        will accept unicode string or will attempt to convert bytestrings
+        into unicode strings using the optional ``encoding`` parameter.
+        ``encoding`` defaults to UTF-8, so no work is needed for bytestrings
+        in UTF-8, but for other character sets like 'cp1252' or 'latin1',
+        ``encoding`` must be set appropriately.
+
+    >>> h = hl7.parse_batch(message)
+
+    To decode a non-UTF-8 byte string::
+
+       hl7.parse_batch(message, encoding='latin1')
+
+    :rtype: :py:class:`hl7.Batch`
+
+    .. [#] http://wiki.hl7.org/index.php?title=Character_Set_used_in_v2_messages
+
+    """
+    # Ensure we are working with unicode data, decode the bytestring
+    # if needed
+    if isinstance(lines, bytes):
+        lines = lines.decode(encoding)
+    batch = None
+    messages = []
+    # Split the batch into lines, retaining the ends
+    for line in lines.strip(hl7_whitespace).splitlines(keepends=True):
+        # strip out all whitespace MINUS the '\r'
+        line = line.strip(hl7_whitespace)
+        if line[:3] == "BHS":
+            if batch:
+                raise ValueError("Batch cannot have more than one BHS segment")
+            batch = line
+        elif line[:3] == "BTS":
+            if not batch:
+                logger.warning("BTS received before BHS: {}".format(line))
+                continue
+            if "\rBTS" in batch:
+                raise ValueError("Batch cannot have more than one BTS segment")
+            batch += line
+        elif line[:3] == "MSH":
+            messages.append(line)
+        else:
+            if not messages:
+                logger.error("Segment received before message header {}".format(line))
```

Comment:
Same comment as before for both the logging statements in this function.  

Logging an error and silently proceeding particularly feels odd, what do you think?

Reply:
I've revisited all of these and updated them to either throw an exception or to ignore the error and drop the offending segment.